[DNM][prototype]: WIP schema changer prototyping

ajwerner opened the pull request.

adityamaru committed:
sql: rewrite index backfiller to stream back progress

Previously, the DistSQL version index backfiller did a a few things
which were non-idiomatic:

- There was fixed time duration and a buffer fill threshold at which the
  backfiller would stop constructing/ingesting index entries and update
the job with its progress. This would potentially cause the DistSQL flow
to get planned, run and tore down multiple times within a single index
backfill schema change.

- The index backfiller relied on the chunk backfiller which performed
  all operations on a fixed "chunk" of rows at a time. Future plans to
make this a memory account/dynamically changing unit of operation meant
that we had to move away from this model.

The new index backfiller logic takes inspiration from our import
coordinator and processors setup. In general, the DistSQL flow is setup
once and the various processors are assigned the spans to backfill.
Every processor concurrently constructs one or more "index entry batch"
and ingests the associated KVs. Periodic go routines stream back the
completed spans to the coordinator node, which in turn periodically
writes the fraction progressed and remaining spans (to account for job
resumption) to the schema change job.

While the size of the index entries in a batch is currently fixed,
future work to be memory aware via the index backfiller mem monitor
should not be particularly difficult to add.

It is important to note that the core index entry building logic has not
been changed. Neither has the column backfiller or the logic for index
backfilling in a txn.

Release note: None

Commit:
Scratch of the beginning of a prototype

Commit:
more prototyping, testing of builder hooked up

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56832)
<!-- Reviewable:end -->

Commit:
implement add column op, schema changer entry point

Commit:
fix up some testing deps to deal with restarts

Commit:
add some notes on the compiler and steps

Commit:
notes on compilation with index backfiller

Commit:
targets, ops, compiler interface

Commit:
sql/schemachange/compiler: add initial graph building code

Commit:
make compile work

Commit:
add a few compiler tests

Commit:
starting to make progress on the executor

Commit:
use {Index,Column}Descriptors to store metadata

Commit:
start of the new builder

Commit:
WIP

Release note (<category, see below>): <what> <show> <why>

Commit:
update dependencies

Commit:
add a modicum of testing to the executor

Commit:
minor refactor to test infra

Commit:
string together an end-to-end test

Commit:
partial implementation of add column in the builder, testing

Commit:
add an end-to-end test of builder

Commit:
hook up session variable for using new schema changer

Commit:
theoretically hook up new schema changer

Commit:
delete old files, generate bazel files

Commit:
make targets into protos

Commit:
start job

Commit:
closer to hooking it up, still have a bug somewhere

Commit:
hook up new schema changer to a job

Commit:
add EXPLAIN (DDL) to show stages and dependencies

Commit:
delete some cruft

Commit:
fix the primary key swap for column changes

Commit:
sql/rowexec: move WriteResumeSpans to testing as it is unused

Release note: None

Commit:
add logic tests for add column, fix a few things

Commit:
real progress on plumbing for index backfiller

Commit:
schemachanger: adopt declarative syntax in compiler

The unit tests in this package don't work but the logictests are working.

Release note: None

Commit:
fix some context plumbing

Commit:
rename the packages

Release note (<category, see below>): <what> <show> <why>

Commit:
fix some bugs

Commit:
WIP

Release note (<category, see below>): <what> <show> <why>

Commit:
refactor/fix computed column validation

Commit:
add datadriven test for builder

Commit:
more WIP

Commit:
sql: add support for computed and default column generation in index backfiller

Release note (<category, see below>): <what> <show> <why>

Commit:
add logictests for index backfill of computed and default columns
